### ISSUE TITLE
bot tries to build wonder stage when all stages are built

### ISSUE THREAD

**Issue by tzyganu** [joffrey-bion/seven-wonders#78]
Steps to reproduce:
 - Play a few times until this happens :smile: It reproduces more often when playing against 6 bots.
 - Bot has all wonder stages built.
 - Bot tries to build a wonder stage
Expected result.
 - Bot should not be allowed to attempt a wonder stage build.
Actual result.
 - game stops.
 - I can only select a card to play but cannot play it.
Additional information.
Logs look like this:
In this case "Player 5 (T-800)" was "the cheater" so there might be some noise in the logs. 
I copy/pasted from the first line that mentioned player 5 for the current round

```
2021-02-02 11:23:42.745  INFO 1 --- [nboundChannel-4] o.l.s.server.controllers.GameController  : Game 1: player 'T-800' (player5) prepared move PlayerMove(type=UPGRADE_WONDER, cardName=Laboratory, transactions=[{2 coin(s) to LEFT_PLAYER for [1 STONE]}, {1 coin(s) to RIGHT_PLAYER for [1 STONE]}])
2021-02-02 11:23:42.795  INFO 1 --- [atcher-worker-2] SevenWondersBot                          : BOT ConnectedPlayer(username=player8, displayName=B-Max, isHuman=false, icon=Icon(name=desktop)): playing turn (action PLAY, 3 cards in hand)
2021-02-02 11:23:42.796  INFO 1 --- [boundChannel-14] o.l.s.server.controllers.GameController  : Game 1: player 'B-Max' (player8) prepared move PlayerMove(type=PLAY, cardName=Caravansery, transactions=[])
2021-02-02 11:23:42.799  INFO 1 --- [atcher-worker-7] SevenWondersBot                          : BOT ConnectedPlayer(username=player7, displayName=HAL, isHuman=false, icon=Icon(name=desktop)): playing turn (action PLAY, 3 cards in hand)
2021-02-02 11:23:42.800 ERROR 1 --- [nboundChannel-1] o.l.s.server.ExceptionHandler            : Uncaught exception thrown during message handling

org.luxons.sevenwonders.engine.moves.InvalidMoveException: Player 5 cannot perform move UPGRADE_WONDER: all levels are already built, or the given resources are insufficient
	at org.luxons.sevenwonders.engine.moves.BuildWonderMove.<init>(BuildWonderMove.kt:15) ~[sw-engine.jar!/:na]
	at org.luxons.sevenwonders.engine.moves.MoveKt.resolve(Move.kt:32) ~[sw-engine.jar!/:na]
	at org.luxons.sevenwonders.engine.Game.prepareMove(Game.kt:118) ~[sw-engine.jar!/:na]
	at org.luxons.sevenwonders.server.controllers.GameController.prepareMove(GameController.kt:89) ~[classes!/:na]
	at jdk.internal.reflect.GeneratedMethodAccessor51.invoke(Unknown Source) ~[na:na]
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[na:na]
	at java.base/java.lang.reflect.Method.invoke(Method.java:564) ~[na:na]
	at org.springframework.messaging.handler.invocation.InvocableHandlerMethod.doInvoke(InvocableHandlerMethod.java:171) ~[spring-messaging-5.3.1.jar!/:5.3.1]
	at org.springframework.messaging.handler.invocation.InvocableHandlerMethod.invoke(InvocableHandlerMethod.java:120) ~[spring-messaging-5.3.1.jar!/:5.3.1]
	at org.springframework.messaging.handler.invocation.AbstractMethodMessageHandler.handleMatch(AbstractMethodMessageHandler.java:565) ~[spring-messaging-5.3.1.jar!/:5.3.1]
	at org.springframework.messaging.simp.annotation.support.SimpAnnotationMethodMessageHandler.handleMatch(SimpAnnotationMethodMessageHandler.java:511) ~[spring-messaging-5.3.1.jar!/:5.3.1]
	at org.springframework.messaging.simp.annotation.support.SimpAnnotationMethodMessageHandler.handleMatch(SimpAnnotationMethodMessageHandler.java:94) ~[spring-messaging-5.3.1.jar!/:5.3.1]
	at org.springframework.messaging.handler.invocation.AbstractMethodMessageHandler.handleMessageInternal(AbstractMethodMessageHandler.java:520) ~[spring-messaging-5.3.1.jar!/:5.3.1]
	at org.springframework.messaging.handler.invocation.AbstractMethodMessageHandler.handleMessage(AbstractMethodMessageHandler.java:454) ~[spring-messaging-5.3.1.jar!/:5.3.1]
	at org.springframework.messaging.support.ExecutorSubscribableChannel$SendTask.run(ExecutorSubscribableChannel.java:144) ~[spring-messaging-5.3.1.jar!/:5.3.1]
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1130) ~[na:na]
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:630) ~[na:na]
	at java.base/java.lang.Thread.run(Thread.java:832) ~[na:na]

```

Screenshots:
 This is what I see
![image](https://user-images.githubusercontent.com/2447784/106594649-10031980-655b-11eb-9bbb-349f0f89cb6d.png)

This is how "player 5"'s board looks like

![image](https://user-images.githubusercontent.com/2447784/106594750-35902300-655b-11eb-87ad-03fe7842c1f5.png)

**Comment by joffrey-bion** [comment]
Thanks for reporting this! I noticed this bug a while ago, and it was (I assumed) due to synchronization problems, which I fixed since then.
I'm surprised to see it again, but will definitely look into this.

Too bad there is no stable way to reproduce this 😢 

Did you run this on your local machine, or did you see this on https://seven-wonders-online.herokuapp.com ?

**Comment by joffrey-bion** [comment]
If you ever run into this again, please take a look at neighbour boards. If this screenshot really is player 5, then he had not built his wonder, but maybe he could just not access the 2 stones required for this.

**Comment by joffrey-bion** [comment]
Reproduced:

```
2021-02-02T12:06:41.608238+00:00 app[web.1]: 2021-02-02 12:06:41.608  INFO 3 --- [atcher-worker-2] SevenWondersBot                          : BOT ConnectedPlayer(username=player9, displayName=T-800, isHuman=false, icon=Icon(name=desktop)): playing turn (action PLAY, 7 cards in hand)
2021-02-02T12:06:41.610706+00:00 app[web.1]: 2021-02-02 12:06:41.610  INFO 3 --- [nboundChannel-2] o.l.s.server.controllers.GameController  : Game 1: player 'T-800' (player9) prepared move PlayerMove(type=UPGRADE_WONDER, cardName=Lighthouse, transactions=[{2 coin(s) to RIGHT_PLAYER for [1 PAPYRUS]}])
2021-02-02T12:06:41.638411+00:00 app[web.1]: 2021-02-02 12:06:41.637  INFO 3 --- [atcher-worker-5] SevenWondersBot                          : BOT ConnectedPlayer(username=player13, displayName=Sonny, isHuman=false, icon=Icon(name=desktop)): playing turn (action PLAY, 7 cards in hand)
2021-02-02T12:06:41.639621+00:00 app[web.1]: 2021-02-02 12:06:41.639  INFO 3 --- [boundChannel-15] o.l.s.server.controllers.GameController  : Game 1: player 'Sonny' (player13) prepared move PlayerMove(type=PLAY, cardName=Lodge, transactions=[{3 coin(s) to RIGHT_PLAYER for [1 CLAY, 1 PAPYRUS]}])
2021-02-02T12:06:41.657688+00:00 app[web.1]: 2021-02-02 12:06:41.657  INFO 3 --- [atcher-worker-2] SevenWondersBot                          : BOT ConnectedPlayer(username=player11, displayName=HAL, isHuman=false, icon=Icon(name=desktop)): playing turn (action PLAY, 7 cards in hand)
2021-02-02T12:06:41.673299+00:00 app[web.1]: 2021-02-02 12:06:41.672 ERROR 3 --- [nboundChannel-5] o.l.s.server.ExceptionHandler            : Uncaught exception thrown during message handling
2021-02-02T12:06:41.673301+00:00 app[web.1]:
2021-02-02T12:06:41.673302+00:00 app[web.1]: org.luxons.sevenwonders.engine.moves.InvalidMoveException: Player 4 cannot perform move PLAY: requirements not met to play the card Gardens
2021-02-02T12:06:41.673302+00:00 app[web.1]: at org.luxons.sevenwonders.engine.moves.PlayCardMove.<init>(PlayCardMove.kt:13) ~[sw-engine.jar!/:na]
2021-02-02T12:06:41.673303+00:00 app[web.1]: at org.luxons.sevenwonders.engine.moves.MoveKt.resolve(Move.kt:29) ~[sw-engine.jar!/:na]
2021-02-02T12:06:41.673304+00:00 app[web.1]: at org.luxons.sevenwonders.engine.Game.prepareMove(Game.kt:118) ~[sw-engine.jar!/:na]
2021-02-02T12:06:41.673304+00:00 app[web.1]: at org.luxons.sevenwonders.server.controllers.GameController.prepareMove(GameController.kt:89) ~[classes!/:na]
2021-02-02T12:06:41.673304+00:00 app[web.1]: at jdk.internal.reflect.GeneratedMethodAccessor50.invoke(Unknown Source) ~[na:na]
2021-02-02T12:06:41.673306+00:00 app[web.1]: at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[na:na]
2021-02-02T12:06:41.673306+00:00 app[web.1]: at java.base/java.lang.reflect.Method.invoke(Method.java:564) ~[na:na]
2021-02-02T12:06:41.673307+00:00 app[web.1]: at org.springframework.messaging.handler.invocation.InvocableHandlerMethod.doInvoke(InvocableHandlerMethod.java:171) ~[spring-messaging-5.3.1.jar!/:5.3.1]
2021-02-02T12:06:41.673307+00:00 app[web.1]: at org.springframework.messaging.handler.invocation.InvocableHandlerMethod.invoke(InvocableHandlerMethod.java:120) ~[spring-messaging-5.3.1.jar!/:5.3.1]
2021-02-02T12:06:41.673308+00:00 app[web.1]: at org.springframework.messaging.handler.invocation.AbstractMethodMessageHandler.handleMatch(AbstractMethodMessageHandler.java:565) ~[spring-messaging-5.3.1.jar!/:5.3.1]
2021-02-02T12:06:41.673309+00:00 app[web.1]: at org.springframework.messaging.simp.annotation.support.SimpAnnotationMethodMessageHandler.handleMatch(SimpAnnotationMethodMessageHandler.java:511) ~[spring-messaging-5.3.1.jar!/:5.3.1]
2021-02-02T12:06:41.673310+00:00 app[web.1]: at org.springframework.messaging.simp.annotation.support.SimpAnnotationMethodMessageHandler.handleMatch(SimpAnnotationMethodMessageHandler.java:94) ~[spring-messaging-5.3.1.jar!/:5.3.1]
2021-02-02T12:06:41.673310+00:00 app[web.1]: at org.springframework.messaging.handler.invocation.AbstractMethodMessageHandler.handleMessageInternal(AbstractMethodMessageHandler.java:520) ~[spring-messaging-5.3.1.jar!/:5.3.1]
2021-02-02T12:06:41.673311+00:00 app[web.1]: at org.springframework.messaging.handler.invocation.AbstractMethodMessageHandler.handleMessage(AbstractMethodMessageHandler.java:454) ~[spring-messaging-5.3.1.jar!/:5.3.1]
2021-02-02T12:06:41.673311+00:00 app[web.1]: at org.springframework.messaging.support.ExecutorSubscribableChannel$SendTask.run(ExecutorSubscribableChannel.java:144) ~[spring-messaging-5.3.1.jar!/:5.3.1]
2021-02-02T12:06:41.673312+00:00 app[web.1]: at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1130) ~[na:na]
2021-02-02T12:06:41.673312+00:00 app[web.1]: at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:630) ~[na:na]
2021-02-02T12:06:41.673313+00:00 app[web.1]: at java.base/java.lang.Thread.run(Thread.java:832) ~[na:na]
2021-02-02T12:06:41.673313+00:00 app[web.1]:
2021-02-02T12:06:41.679124+00:00 app[web.1]: 2021-02-02 12:06:41.679  INFO 3 --- [atcher-worker-5] SevenWondersBot                          : BOT ConnectedPlayer(username=player10, displayName=Wall-E, isHuman=false, icon=Icon(name=desktop)): playing turn (action PLAY, 7 cards in hand)
2021-02-02T12:06:41.681570+00:00 app[web.1]: 2021-02-02 12:06:41.681  INFO 3 --- [boundChannel-16] o.l.s.server.controllers.GameController  : Game 1: player 'Wall-E' (player10) prepared move PlayerMove(type=UPGRADE_WONDER, cardName=Palace, transactions=[{2 coin(s) to LEFT_PLAYER for [1 STONE]}])
2021-02-02T12:06:41.710590+00:00 app[web.1]: 2021-02-02 12:06:41.710  INFO 3 --- [atcher-worker-2] SevenWondersBot                          : BOT ConnectedPlayer(username=player12, displayName=GLaDOS, isHuman=false, icon=Icon(name=desktop)): playing turn (action PLAY, 7 cards in hand)
2021-02-02T12:06:41.712370+00:00 app[web.1]: 2021-02-02 12:06:41.711  INFO 3 --- [boundChannel-13] o.l.s.server.controllers.GameController  : Game 1: player 'GLaDOS' (player12) prepared move PlayerMove(type=PLAY, cardName=Observatory, transactions=[])
2021-02-02T12:06:41.751154+00:00 app[web.1]: 2021-02-02 12:06:41.750  INFO 3 --- [atcher-worker-5] SevenWondersBot                          : BOT ConnectedPlayer(username=player8, displayName=B-Max, isHuman=false, icon=Icon(name=desktop)): playing turn (action PLAY, 7 cards in hand)
2021-02-02T12:06:41.752878+00:00 app[web.1]: 2021-02-02 12:06:41.752  INFO 3 --- [boundChannel-12] o.l.s.server.controllers.GameController  : Game 1: player 'B-Max' (player8) prepared move PlayerMove(type=PLAY, cardName=Gardens, transactions=[{2 coin(s) to LEFT_PLAYER for [1 CLAY]}])
```

Boards:
![image](https://user-images.githubusercontent.com/1913296/106598630-4a1eec00-6558-11eb-8dd6-e638c722ee6a.png)
![image](https://user-images.githubusercontent.com/1913296/106598657-530fbd80-6558-11eb-8cb9-0b3626af70ad.png)
![image](https://user-images.githubusercontent.com/1913296/106598679-599e3500-6558-11eb-803c-08673af249ba.png)

**Comment by tzyganu** [comment]
Oh, you move to fast with the issue solving :)

> Did you run this on your local machine, or did you see this on https://seven-wonders-online.herokuapp.com ?
I reproduced this locally

> You said the bot had all levels built, but in the screenshot it doesn't look like so. Which of your statements is correct? :smile: 
I have no idea which is correct. I sent the logs and the screenshot. I have no idea which is actually true. Apparently both :smile:

**Comment by joffrey-bion** [comment]
I think I found the problem. Since the addition of this feature: https://github.com/joffrey-bion/seven-wonders/issues/50, the bots can now see *all* transactions (not only the cheapest ones).

The problem is that there is no proper filtering of the transactions, we just check whether the player has enough gold for the cheapest one. Definitely a bug that also affects human players.

**Comment by tzyganu** [comment]
OK will close this one as duplicate of or at least included in #81
